[Merged by Bors] - feat: simps uses fields of parent structures

fpvandoorn opened the pull request:
* `initialize_simps_projections` now by default generates all projections of all parent structures, and doesn't generate the projections to those parent structures.
* You can also rename a nested projection directly, without having to specify intermediate parent structures
* Added the option to turn the default behavior off (done in e.g. `TwoPointed`)

Internal changes:

* Move most declarations to the Simps namespace, and shorten their names
* Restructure `ParsedProjectionData` to avoid the bug [reported here](https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/allocation.20crash.3F) (and to another bug where it seemed that the wrong data was inserted in `ParsedProjectionData`, but it was hard to minimize because of all the crashes). If we manage to fix the bug in that Zulip thread, I'll see if I can track down the other bug in commit 97454284bb94b8aa45df96316731bdf2c04fc3ce

---

This likely conflicts with future PRs, so please delegate instead of merge

<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

gebner commented on Mathlib/Tactic/Simps/Basic.lean:
```diff
@@ -445,61 +446,104 @@ def projectionsInfo (l : List ProjectionData) (pref : String) (str : Name) : Mes
+Note: `$e.$projName` is not applied to further (implicit) arguments. -/
+def elabProjectionNotation (univs : List Name) (e : Expr) (strName projName : Name) :
+  MetaM Expr :=
+TermElabM.run' (s := {levelNames := univs}) <| do
```
```suggestion
TermElabM.run' (s := {levelNames := univs}) do
```
Oof, this isn't pretty.  I think it would be nicer to stay in `MetaM` entirely, and avoid accessing all these elaborator internals.

Building a non-subobject projection by hand shouldn't be too hard, I think we can just copy the params+univ params:
```lean
def mkDirectProjection (e : Expr) (fieldName : Name) : MetaM Expr := do
  let ty ← whnf (←inferType e)
  let .const structName us := ty.getAppFn | ...
  let some projName := getProjFnForField? (← getEnv) structName fieldName | ...
  -- also check out Meta.mkProjFn
  return mkAppN (.const projName us) (ty.getAppArgs.push e)
```

and then we could extend that to subobjects pretty easily:
```lean
def mkProjection (e : Expr) (fieldName : Name) : MetaM Expr := do
  let .const structName _ := (← whnf (←inferType e)).getAppFn | ..
  let some baseStruct := findField? (← getEnv) structName fieldName | ..
  let mut e := e
  for subobj in (getPathToBaseStructure? (← getEnv) baseStruct structName).get! do
    e ← mkDirectProjection e subobj
  mkDirectProjection e fieldName
```

fpvandoorn replied:
I was doing it manually before, and I got the remark that I was re-implementing parts of the elaborator, and it was maybe better to reuse it. But I'm happy to do it manually again.

fpvandoorn commented:
I separated out roughly the definitions you wrote. Thanks for that, they were much nicer than the versions I had myself earlier.
I added some additional changes to this PR:
* I had to change `mkParsedProjectionData`, to put parts of it in the shape `@[simps]` expects (and documented these requirements)
* I changed the rules for `initialize_simps_projections` so that the default behavior is better (that was already part of the previous review), but for the sake of mathlib3-backwards compatibility I added the option to turn the default behavior off and revert to the previous behavior (done in e.g. `TwoPointed`)
* I restructured `ParsedProjectionData`, otherwise I would often run into the bug [reported here](https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/allocation.20crash.3F) (and to another bug where it seemed that the wrong data was inserted in `ParsedProjectionData`, but it was hard to minimize because of all the crashes).

kmill commented:
It seems like an improvement to me, and given there haven't been any more comments in the last few weeks:

bors d+

fpvandoorn commented:
bors merge
